samples/msal-node-samples/on-behalf-of-distributed-cache/src/AuthProvider.ts

```diff
@@ -0,0 +1,154 @@
+    AuthenticationResult,
+    CryptoProvider,
+    OnBehalfOfRequest
+} from "@azure/msal-node";
+
+import CustomCachePlugin from "./CustomCachePlugin";
+import RedisClientWrapper from "./RedisClientWrapper";
+import AxiosHelper from './AxiosHelper';
+
+export type AppConfig = {
+    instance: string;
+    tenantId: string;
+    clientId: string;
+    clientSecret: string;
+    permissions: string;
+};
+
+export class AuthProvider {
+    private msalConfig: Configuration;
+    private cryptoProvider: CryptoProvider;
+    private cacheClientWrapper: RedisClientWrapper;
+
+    private constructor(msalConfig: Configuration, cacheClient: RedisClientType) {
+        this.msalConfig = msalConfig;
+        this.cryptoProvider = new CryptoProvider();
+        this.cacheClientWrapper = new RedisClientWrapper(cacheClient);
+    }
+
+    static async initialize(appConfig: AppConfig, cacheClient: RedisClientType): Promise<AuthProvider> {
+        const msalConfig = {
+            auth: {
+                clientId: appConfig.clientId,
+                authority: `${appConfig.instance}/${appConfig.tenantId}`,
+                clientSecret: appConfig.clientSecret,
+            },
+            system: {
+                loggerOptions: {
+                    loggerCallback(loglevel, message, containsPii) {
+                        console.log(message);
+                    },
+                    logLevel: LogLevel.Trace,
+                    piiLoggingEnabled: false
+                },
+                // proxyUrl: "http://localhost:8888" // uncomment to capture traffic with Fiddler
+            }
+        } as Configuration;
+
+        const msalConfigWithMetadata = await AuthProvider.getMetadata(msalConfig, cacheClient);
+
+        return new AuthProvider(msalConfigWithMetadata, cacheClient);
+    }
+
+    async getToken(tokenRequest: OnBehalfOfRequest): Promise<AuthenticationResult | null> {
+        const partitionKey = await this.cryptoProvider.hashString(tokenRequest.oboAssertion);
+
+        const cca = new ConfidentialClientApplication({
+            ...this.msalConfig,
+            cache: {
+                cachePlugin: new CustomCachePlugin(
+                    this.cacheClientWrapper,
+                    partitionKey // partitionKey hash(oboAssertion)
+                )
+            }
+        });
+
+        let tokenResponse = null;
+
+        try {
+            performance.mark("acquireTokenOnBehalfOf-start");
+            await cca.getTokenCache().getAllAccounts(); // required for cache read
+            tokenResponse = await cca.acquireTokenOnBehalfOf(tokenRequest);
+            performance.mark("acquireTokenOnBehalfOf-end");
+            performance.measure(
+                tokenResponse?.fromCache ? "acquireTokenOnBehalfOf-fromCache" : "acquireTokenOnBehalfOf-fromNetwork",
+                "acquireTokenOnBehalfOf-start",
+                "acquireTokenOnBehalfOf-end"
+            );
+        } catch (error) {
```

Comment:
catch / throw is not needed.

Reply:
My bad -it's not doing anything here: errors thrown here will be caught in the call site already (app.ts)